### PR TITLE
Add Magpie-TTS and Updates NeMo Audio Codecs

### PR DESCRIPTION
**Pull request by paarthneekhara**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/tts/models/t5tts.py [hunk 1]
```diff
@@ -0,0 +1,1528 @@
...
+    )
+    dataset.text_tokenizer = tokenizer
+    dataset.text_conditioning_tokenizer = text_conditioning_tokenizer
+
+
+class T5TTS_Model(ModelPT):
+    def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
+        # Convert to Hydra 1.0 compatible DictConfig
+        cfg = model_utils.convert_model_config_to_dict_config(cfg)
+        cfg = model_utils.maybe_update_config_version(cfg)
```
Comment (XuesongYang) [inline nemo/collections/tts/models/t5tts.py:101]: we should remove these two lines to take care of any legacy hydra<1.0 since our recipe starts with `hydra-core>1.3,<=1.3.2`

File: nemo/collections/tts/modules/audio_codec_modules.py [hunk 2]
```diff
@@ -167,6 +174,318 @@ def forward(self, audio_real, audio_gen):
...
+    def __init__(self, coefficient=0.97):
+        super().__init__()
+        self.coefficient = coefficient
+        self.register_buffer("filter", torch.FloatTensor([-self.coefficient, 1.0]).unsqueeze(0).unsqueeze(0))
+
+    def forward(self, x):
+        assert len(x.size()) == 2
+
+        x = torch.nn.functional.pad(x.unsqueeze(1), (1, 0), "reflect")
+        return torch.nn.functional.conv1d(x, self.filter).squeeze(1)
```
Comment (blisc) [inline nemo/collections/tts/modules/audio_codec_modules.py:248]: What's the difference between this and https://github.com/NVIDIA/NeMo/blob/e5899c37a27952bf003c2b8d082ebee68fbb4f7f/nemo/collections/asr/parts/preprocessing/features.py#L433

File: scripts/t5tts/dpo/create_text_contextpairs.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+
+def write_manifest(fp, records):
+    with open(fp, "w") as f:
+        for record in records:
+            f.write(json.dumps(record) + "\n")
+    print("Wrote {} records to: {}".format(len(records), fp))
+
+
+def main():
```
Comment (blisc) [inline scripts/t5tts/dpo/create_text_contextpairs.py:13]: Can we document this script? What does it do and what are the args?

File: scripts/t5tts/dpo/create_preference_pairs.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+import copy
+import json
+import math
+import os
+import random
+
+from tqdm import tqdm
+
+
+def main():
```
Comment (blisc) [inline scripts/t5tts/dpo/create_preference_pairs.py:11]: Can we document this script? What does it do and what are the args?

File: scripts/t5tts/dpo/create_preference_pairs.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--input_manifest", type=str, default="/Data/testing_240subset.json")
+    parser.add_argument(
+        "--generated_audio_dir",
+        type=str,
+        default="/Data/Experiments/NewT5TTSDPO/Debug/Generations/T5TTS/version_0/audios",
+    )
```
Comment (blisc) [inline scripts/t5tts/dpo/create_preference_pairs.py:18]: Can you remove the defaults?

File: nemo/collections/tts/models/t5tts.py [hunk 6]
```diff
@@ -0,0 +1,1528 @@
...
+
+    if use_text_conditioning_tokenizer:
+        # TODO: make this configurable
+        # Conditioning text tokenizer
+        text_conditioning_tokenizer = T5Tokenizer.from_pretrained("google-t5/t5-small")
+
+    return aggregated_tokenizer, text_conditioning_tokenizer
+
+
+def worker_init_fn(worker_id):
```
Comment (blisc) [inline nemo/collections/tts/models/t5tts.py:84]: What is this function for?

File: nemo/collections/tts/models/t5tts.py [hunk 7]
```diff
@@ -0,0 +1,1528 @@
...
+
+    def setup_test_data(self, cfg):
+        self._test_dl = self._setup_test_dataloader(cfg)
+
+    @classmethod
+    def list_available_models(cls) -> List[PretrainedModelInfo]:
+        return []
+
+
+class T5TTS_ModelInference(T5TTS_Model):
```
Comment (blisc) [inline nemo/collections/tts/models/t5tts.py:1091]: Should we keep this class in the final release or should we keep it for our internal use cases?

**Review by blisc — CHANGES_REQUESTED** [review]

File: nemo/collections/tts/modules/audio_codec_modules.py (hunk 2 above)
Comment (Edresson) [reply nemo/collections/tts/modules/audio_codec_modules.py:248]: They are similar, but the current implementation in NeMo preserves the first value of each vector, whereas the implementation in this class modifies the entire input. Since the Speaker Encoder model was trained using this new implementation, I don’t think we can replace it.

**Review by blisc — DISMISSED** [review]
